### DISCUSSION TITLE
How to improve debug build time for Android?

### DISCUSSION THREAD

**Discussion by DoubleDBE** [AvaloniaUI/Avalonia#14401] [Q&A]
Is there a way to improve build time when debugging for Android?

With every little change, VS performs a complete rebuild, which takes ages before debugging is possible.
Not to mention when you need to debug on an actual (low end) device.

I'm not sure if i'm doing something wrong, or is this something to be improved?

I'm coming from MAUI and had no issues regarding build time. Not saying it was perfect but they made it smart enough to avoid a complete rebuild for every little change.

**Comment by maxkatz6** [comment]
> I'm coming from MAUI and had no issues regarding build time

It shouldn't be much different if different at all.

Please use [MSBuild Structured Log Viewer](https://msbuildlog.com/) app to see what specific build tasks taking the most time.

**Comment by DoubleDBE** [comment]
Thanks for this awesome tool.

It helped me track where most time is spent, but I have no clue how I can solve it.

I have a project in my solution that references quite a few Android Binding nugets which contain .jar and .aar files.

Seems like a lot of time is spent there.

D8 = 8.714 s, 1 calls.
BuildApk = 4.269 s, 1 calls.
CompileAvaloniaXamlTask = 2.708 s, 6 calls.
Javac = 2.214 s, 1 calls.
LinkAssembliesNoShrink = 1.846 s, 1 calls.
ManifestMerger = 951 ms, 1 calls.
AndroidApkSigner = 792 ms, 1 calls.
Csc = 694 ms, 5 calls.
GenerateJavaStubs = 648 ms, 1 calls.
ResolveLibraryProjectImports = 604 ms, 1 calls.

**Reply by maxkatz6** [reply]
Looks like Avalonia only adds 3s to the total build time (CompileAvaloniaXamlTask).
You can try to ask question at https://github.com/xamarin/xamarin-android/pulls as well.

**Reply by maxkatz6** [reply]
Though, 3s is also more than I would expect. But I guess you have many projects with XAML? See "6 calls".

**Comment by DoubleDBE** [comment]
Seems like this solved itself somehow. Not sure how this came to be but whatever.
